### ISSUE TITLE
Splitting do not occur enough

### ISSUE THREAD

**Issue by GriesserP** [BGU-CS-VIL/DeepDPM#18]
Hello thank you very much for your code.

I am trying to use it to cluster text. I am using SBert (all-MiniLM-L6-v2) embeddings of dimension 384 (reduced to dimension 20 with UMAP). I am trying to cluster the Stack-Overflow dataset containing 20 clusters, 1000 datapoints per class (c.f. the scatter plot, where I reduce the embeddings to dim=2 with UMAP). My problem is that without changing the parameters of your model (except for prior_nu which need to be at least codes_dim + 1. so I set it to 22 and also I changed latent_dim to be 20) then the model won't split enough the clusters and is stuck at 1 or 2 clusters. I tried to change the "start_sub_clustering" ,"start_splitting",  "start_merging", "split_every_n_epochs", "merge_every_n_epochs", "split_merge_every_n_epochs" to smaller values but it does not help.

Also because there are a lot of hyperparameters, I am not sure where to start then tuning for the case of text. Do you have any idea or hints to apply your model to such types of embeddings?

![Screenshot 2022-05-19 at 09 21 27](https://user-images.githubusercontent.com/36691390/169235389-ac4d51a1-c2cf-49da-9763-d1956bd9c902.png)
s

**Comment by freifeld** [comment]
Thank you for the question and interest in our paper/code. 
Meitar will probably answer in more detail later, but if you find the default config doesn't give you enough splits then the quickest way to increase the number of splits is to:
1) increase alpha (e.g., alpha=1000, or 10,000)
and/or
2) Pick Psi to be small (e.g., take the default Psi and scale it by a small number) AND set nu to be very high (using a small Psi without a large nu will not help). The reason is that this will mean you are condifent that you want samll clusters, so the model will be driven to use many of them in order to fully cover all the data.  

I would go with option number 2 first and keep alpha as is, but you may want to play with both options.

**Comment by GriesserP** [comment]
Thank you very much for the quick answer. I will try this! By psi you mean the factor that multiply the identiy matrix ? Is it the "cov_const" parameter in your code ? and nu is "prior_nu" parameter right ?

**Comment by GriesserP** [comment]
Thank you for the precision, I will try it out and definitively keep you updated about the results!

**Comment by GriesserP** [comment]
Hello again,
Now the splitting occur, but for the stack-overflow the results aren't very satisfying. For example using kmeans (with the appropriate number of cluster) on top of SBERT embeddings, I get an accuracy of 0.71. Now using DeepDPM, the model finds 15 clusters (which is cool) but when plotting the confusion matrix (after linear assignement) this is what I get:

![Screenshot 2022-05-25 at 09 28 43](https://user-images.githubusercontent.com/36691390/170205638-aaa98924-22e2-41db-9d97-b88315bdfb5a.png)

I also tested on two other dataset and the same problem occur, i.e. each cluster contains approximatively the same amount of datapoint for each true class.
Do you know where this could come from ?

**Comment by abdallah197** [comment]
Hey @GriesserP, can you share the class for the dataset

**Comment by GriesserP** [comment]
Hello @abdallah197, I didn't create a class for the dataset, I only loaded the text and embedded each datapoint using the sentence_transoformer library and passed the data matrix to umap. I saved the embedding and the labels inside the "pretrained embeddings". When calling the main from deepdpm I just passed the path to the folder containing the embeddings, hope it helps

**Comment by lala-ai-dot** [comment]
@GriesserP Hello, may I ask which parameters you adjusted to change the number of clusters? Looking forward to your reply.

> 大家好！ 现在模型已经可以进行拆分了，但 Stack Overflow 上的结果并不令人满意。例如，在 SBERT 嵌入上使用 kmeans（选择合适的聚类数量）时，我的准确率只有 0.71。现在使用 DeepDPM，模型找到了 15 个聚类（这很酷），但在绘制混淆矩阵（在所有训练数据上进行线性赋值后）时，我得到了以下结果：
> 
> ![Screenshot 2022-05-25 at 09 28 43](https://user-images.githubusercontent.com/36691390/170205638-aaa98924-22e2-41db-9d97-b88315bdfb5a.png)
> 
> 我也在另外两个数据集上进行了测试，出现了同样的问题，即每个聚类对于每个真实类包含大约相同数量的数据点。
> 
> 更有趣的是，当我绘制混淆矩阵时，除了验证数据（每个类别 200 个样本）之外，我得到了更好的性能（就好像验证集发生了过拟合一样）。这很奇怪，因为验证数据不应该在训练中使用，
> 
> ![Screenshot 2022-05-25 at 09 39 50](https://user-images.githubusercontent.com/36691390/170207911-4e11a442-3d40-4b0a-bdfc-c4d4ccb23c12.png)
> 
> 你知道这是从哪里来的吗？
> 
> [更新] 抱歉，我在绘制混淆矩阵时犯了一个错误。所有数据的结果都非常令人满意。再次感谢您的贡献，很高兴看到它很好地适用于文本数据。

Hello, may I ask which parameters you adjusted to change the number of clusters? Looking forward to your reply.
